# Sentiment Analysis: Classical ML vs Transformer Models

This project compares a classical machine learning model (Logistic Regression) with a transformer-based model (BERT).

In [6]:
# Install libraries
!pip install transformers datasets scikit-learn

## Method

We use two approaches:

1. Logistic Regression with TF-IDF features  
2. Transformer model (DistilBERT)

The goal is to compare classical and modern NLP methods.

In [11]:


# Import libraries
from datasets import load_dataset
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from transformers import pipeline

# Load dataset
dataset = load_dataset("imdb")
train_data = dataset["train"].shuffle(seed=42).select(range(2000))
train_texts = train_data["text"]
train_labels = train_data["label"]


test_data = dataset["test"].shuffle(seed=42).select(range(500))
test_texts = test_data["text"]
test_labels = test_data["label"]


# Classical ML model
vectorizer = TfidfVectorizer(max_features=5000)
X_train = vectorizer.fit_transform(train_texts)
X_test = vectorizer.transform(test_texts)

model = LogisticRegression()
model.fit(X_train, train_labels)

preds = model.predict(X_test)
acc = accuracy_score(test_labels, preds)
print(acc)

# Transformer model
classifier = pipeline("sentiment-analysis")

sample = test_texts[:10]
results = classifier(sample)

print("Transformer Predictions:")
for r in results:
    print(r)

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f.
Using a pipeline without specifying a model name and revision in production is not recommended.


0.822


Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

Transformer Predictions:
{'label': 'POSITIVE', 'score': 0.9988757967948914}
{'label': 'POSITIVE', 'score': 0.9969831109046936}
{'label': 'NEGATIVE', 'score': 0.9972442388534546}
{'label': 'NEGATIVE', 'score': 0.649212658405304}
{'label': 'NEGATIVE', 'score': 0.9985034465789795}
{'label': 'POSITIVE', 'score': 0.999742329120636}
{'label': 'POSITIVE', 'score': 0.9968295693397522}
{'label': 'NEGATIVE', 'score': 0.9995834231376648}
{'label': 'NEGATIVE', 'score': 0.9980700612068176}
{'label': 'POSITIVE', 'score': 0.999727189540863}


## Results

- Logistic Regression achieved an accuracy of 0.822 on the test set.
- The transformer model produced highly confident sentiment predictions.
## Interpretation

Transformer models capture deeper language patterns compared to classical methods.

## Limitations

- Small subset of dataset used  
- No hyperparameter tuning  
- Transformer model not fine-tuned  

## Future Work

- Fine-tune transformer models  
- Use larger datasets  